### ISSUE TITLE
ES 2.3.3 geo_polygon query across the prime meridian cannot find points near the prime meridian

### ISSUE THREAD

**Issue by laoyaosniper** [elastic/elasticsearch#19900]
**Elasticsearch version**: 2.3.3

**Plugins installed**: nothing

**JVM version**: 1.8.0_77

**OS version**: OS X El Captain 10.11.5 

**Description of the problem including expected versus actual behavior**:
I found a bug when testing geo polygon search across the prime meridian (the 0 longitude). It cannot find the geo point inside the polygon whose longitude is smaller than 1.4 on the eastern hemisphere (positive longitude). Also, it's a regression because those points can be found on ES 1.4.4.

**Steps to reproduce**:
1. Create an index with mapping:
   
   ```
   PUT /geo_polygon
   {
       "settings" : {
           "index" : {
               "number_of_shards" : 1, 
               "number_of_replicas" : 0
           }
       },
   
     "mappings": {
       "my_type": {
         "properties": {
           "location": {
             "type": "geo_point"
           }
         }
       }
     }
   }
   ```
2. Put a list of points on the map. I put all the points on the north hemisphere to eliminate the possible affect from latitude. 
   
   ```
   PUT /geo_polygon/my_type/1
   {
     "text": "Point on the prime meridian",
     "location": { 
       "lat": 30,
       "lon": 0
     }
   }
   
   PUT /geo_polygon/my_type/2
   {
     "text": "Point very close to the prime meridian, eastern hemisphere",
     "location": { 
       "lat": 30,
       "lon": 0.1
     }
   }
   
   PUT /geo_polygon/my_type/3
   {
     "text": "Point very close to the prime meridian, western hemisphere",
     "location": { 
       "lat": 30,
       "lon": -0.1
     }
   }
   
   PUT /geo_polygon/my_type/4
   {
     "text": "Point close to the prime meridian, eastern hemisphere",
     "location": { 
       "lat": 30,
       "lon": 1.4
     }
   }
   
   PUT /geo_polygon/my_type/5
   {
     "text": "Point close to the prime meridian, western hemisphere",
     "location": { 
       "lat": 30,
       "lon": -1.4
     }
   }
   
   PUT /geo_polygon/my_type/6
   {
     "text": "Point close to the prime meridian on eastern hemisphere, different latitude",
     "location": { 
       "lat": 25,
       "lon": 1.4
     }
   }
   
   PUT /geo_polygon/my_type/7
   {
     "text": "Point close to the prime meridian on western hemisphere, different latitude",
     "location": { 
       "lat": 25,
       "lon": -1.4
     }
   }
   
   PUT /geo_polygon/my_type/8
   {
     "text": "Point not too far from the prime meridian",
     "location": { 
       "lat": 30,
       "lon": 1.5
     }
   }
   
   PUT /geo_polygon/my_type/9
   {
     "text": "Point far from the prime meridian",
     "location": { 
       "lat": 30,
       "lon": 20
     }
   }
   
   PUT /geo_polygon/my_type/10
   {
     "text": "Point quite far from the prime meridian",
     "location": { 
       "lat": 30,
       "lon": 90
     }
   }
   ```
3. Search in the following polygon:
   
   ```
   GET /geo_polygon
   {
       "query": {
           "bool" : {
               "filter" : {
                   "geo_polygon" : {
                       "location" : {
                           "points" : [
                               {"lat" : 45, "lon" : 45},
                               {"lat" : 45, "lon" : -45},
                               {"lat" : 10, "lon" : 0}
                           ]
                       }
                   }
               }
           }
       }
   }
   ```

Expected: points 1~9 should be found. The very far point 10 shouldn't be returned.
Actual: only point 3, 5, 7, 8, 9 are found, which have the following characteristic:
1. On the western hemisphere (lon < 0)
2. Or, the longitude is larger than 1.5.

Tried the same thing on ES 1.4 cluster and it could find point 1~9 correctly. 

Currently I could mitigate this problem by splitting the polygon along the prime meridian and do a conjunctive search like the request below. It could find point 1~9 on ES 2.3.3 stack.

``````
```
{
    "query": {
        "bool" : {
            "should" : [
                {
                    "bool": {
                        "filter" : {
                            "geo_polygon" : {
                                "location" : {
                                    "points" : [
                                        {"lat" : 45, "lon" : 45},
                                        {"lat" : 45, "lon" : 0},
                                        {"lat" : 10, "lon" : 0}
                                    ]
                                }
                            }
                        }
                    }
                },
                {
                    "bool": {
                        "filter" : {
                            "geo_polygon" : {
                                "location" : {
                                    "points" : [
                                        {"lat" : 45, "lon" : 0},
                                        {"lat" : 45, "lon" : -45},
                                        {"lat" : 10, "lon" : 0}
                                    ]
                                }
                            }
                        }
                    }
                }
            ]
        }
    }
}
```
``````

Interestingly, I tried to mirror the polygon above to the south hemisphere and search in the mirrored hemisphere. Instead of find nothing, I found points 1, 2, 4, 6, 8 and 9, which are all the points on non-western hemisphere.

**Comment by clintongormley** [comment]
@nknize i can reproduce this in 2.3.5, but fixed in 5.0.0-alpha5.  Could you take a look please?

**Comment by nknize** [comment]
Looks like a bug in the ShapeBuilder that was fixed during the 5.0 query refactoring. I'll dig.

**Comment by imotov** [comment]
This no longer reproduces in 5.x and above and a lot of code changed around it since then. So, I think it is safe to close.
